source/extensions/filters/network/common/redis/client.h

```diff
@@ -203,13 +203,72 @@ class ClientFactory {
    * @param redis_command_stats supplies the redis command stats.
    * @param scope supplies the stats scope.
    * @param auth password for upstream host.
+   * @param is_transaction_client true if this client was created to relay a transaction.
    * @return ClientPtr a new connection pool client.
    */
   virtual ClientPtr create(Upstream::HostConstSharedPtr host, Event::Dispatcher& dispatcher,
                            const Config& config,
                            const RedisCommandStatsSharedPtr& redis_command_stats,
                            Stats::Scope& scope, const std::string& auth_username,
-                           const std::string& auth_password) PURE;
+                           const std::string& auth_password, bool is_transaction_client) PURE;
+};
+
+// A MULTI command sent when starting a transaction.
+struct MultiRequest : public Extensions::NetworkFilters::Common::Redis::RespValue {
+public:
+  MultiRequest() {
+    type(Extensions::NetworkFilters::Common::Redis::RespType::Array);
+    std::vector<NetworkFilters::Common::Redis::RespValue> values(1);
+    values[0].type(NetworkFilters::Common::Redis::RespType::BulkString);
+    values[0].asString() = "MULTI";
+    asArray().swap(values);
+  }
+};
+
+// An empty array sent when a transaction is empty.
+struct EmptyArray : public Extensions::NetworkFilters::Common::Redis::RespValue {
+public:
+  EmptyArray() {
+    type(Extensions::NetworkFilters::Common::Redis::RespType::Array);
+    std::vector<NetworkFilters::Common::Redis::RespValue> values;
+    asArray().swap(values);
+  }
+};
+
+// A class representing a Redis transaction.
+class Transaction {
+public:
+  Transaction(Network::ConnectionCallbacks* connection_cb) : connection_cb_(connection_cb) {}
+  ~Transaction() {
+    if (connection_established_) {
+      client_->close();
+      connection_established_ = false;
+    }
+  }
+
+  void start() { active_ = true; }
+
+  void close() {
+    active_ = false;
+    key_.clear();
+    if (connection_established_) {
+      client_->close();
+      connection_established_ = false;
+    }
+    should_close_ = false;
+  }
+
+  bool active_{false};
+  bool connection_established_{false};
+  bool should_close_{false};
+  std::string key_;
+  ClientPtr client_;
+  Network::ConnectionCallbacks* connection_cb_;
```

Comment:
Could the variables be made private?

Reply:
I thought this might be a bit redundant since I will have to add a getter and a setter for each var, so I decided to treat it more like a struct. What do you think?

Reply:
Hmm.. I think Envoy code style seems to make variables private in a class. Maybe you could simply make it a struct?